### PR TITLE
Implement attributes for the `<meter>` element

### PR DESCRIPTION
**Pull request by shanehandley**
Implement attributes for the html `<meter>` element.

https://html.spec.whatwg.org/multipage/form-elements.html#the-meter-element

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

File: components/script/dom/htmlmeterelement.rs [hunk 1]
```diff
@@ -52,4 +57,138 @@ impl HTMLMeterElement {
...
+        let max = *self.Max();
+
+        Finite::wrap(
+            self.upcast::<Element>()
+                .get_string_attribute(&local_name!("value"))
+                .parse_floating_point_number()
+                .map_or(0.0, |candidate_actual_value| {
+                    if candidate_actual_value < min || min > max {
+                        return min;
+                    }
```
Comment (sagudev) [inline components/script/dom/htmlmeterelement.rs:73]: Why is this needed? Isn't clamp enough?

File: components/script/dom/htmlmeterelement.rs [hunk 2]
```diff
@@ -52,4 +57,138 @@ impl HTMLMeterElement {
...
+        let max = *self.Max();
+
+        Finite::wrap(
+            self.upcast::<Element>()
+                .get_string_attribute(&local_name!("low"))
+                .parse_floating_point_number()
+                .map_or(min, |candidate_low_boundary| {
+                    if candidate_low_boundary < min || min > max {
+                        return min;
+                    }
```
Comment (sagudev) [inline components/script/dom/htmlmeterelement.rs:131]: ditto

File: components/script/dom/htmlmeterelement.rs [hunk 3]
```diff
@@ -52,4 +57,138 @@ impl HTMLMeterElement {
...
+        let low: f64 = *self.Low();
+
+        Finite::wrap(
+            self.upcast::<Element>()
+                .get_string_attribute(&local_name!("high"))
+                .parse_floating_point_number()
+                .map_or(max, |candidate_high_boundary| {
+                    if candidate_high_boundary < low || min > max {
+                        return low;
+                    }
```
Comment (sagudev) [inline components/script/dom/htmlmeterelement.rs:157]: ditto

File: components/script/dom/htmlmeterelement.rs [hunk 4]
```diff
@@ -52,4 +57,138 @@ impl HTMLMeterElement {
...
+        let min = *self.Min();
+
+        Finite::wrap(
+            self.upcast::<Element>()
+                .get_string_attribute(&local_name!("optimum"))
+                .parse_floating_point_number()
+                .map_or(max.add(min).div(2.0), |candidate_optimum_point| {
+                    if min <= max {
+                        return candidate_optimum_point.clamp(min, max);
+                    }
```
Comment (sagudev) [inline components/script/dom/htmlmeterelement.rs:182]: ditto

File: tests/wpt/meta/html/dom/reflection-forms.html.ini [hunk 5]
```diff
@@ -3380,132 +3380,24 @@
...
-    expected: FAIL
-
-  [meter.value: IDL set to -10000000000]
-    expected: FAIL
-
-  [meter.value: IDL set to -1]
-    expected: FAIL
-
   [meter.value: IDL set to -0]
     expected: FAIL
```
Comment (sagudev) [inline tests/wpt/meta/html/dom/reflection-forms.html.ini:3384]: Apparently some setters do not work correctly (maybe wrong parsing?).

File: components/script/dom/htmlmeterelement.rs (hunk 1 above)
Comment (shanehandley) [reply components/script/dom/htmlmeterelement.rs]: Oh avoiding panics while working on WPT passes, I can't reproduce those cases now, so removed 👍

File: tests/wpt/meta/html/dom/reflection-forms.html.ini (hunk 5 above)
Comment (shanehandley) [reply tests/wpt/meta/html/dom/reflection-forms.html.ini:3384]: Sorry for not responding to this comment yet but I went down a rabbit hole, I switched the setter implementations to `parse_floating_point_number` before setting the value, but the issue persisted.

I'm really unsure if this is something in `parse_floating_point_number` or not :/

[Spidermonkey's](https://searchfox.org/mozilla-central/source/dom/base/ResponsiveImageSelector.cpp#36) `parseFloat(-0.0)` returns `0`. But our implementation returns -0.0 - do you think this might be a factor or am I barking up the wring tree?

Thank you!

Comment (sagudev) [reply tests/wpt/meta/html/dom/reflection-forms.html.ini:3384]: I think that our parse_floating_point_number is wrong, but we can leave this for follow up. In firefox the corresponding function is actually https://searchfox.org/mozilla-central/source/dom/base/nsAttrValue.cpp#1892 (you can see it's only used on progress and meter element just like ours).

**Review by sagudev — APPROVED** [review]

**Comment by sagudev** [comment]
Thank you!
